### PR TITLE
Refactor: change source test listener to expect certain events

### PR DESCRIPTION
**Pull request by KauzClay**
## Why
Currently, in the source_test.go test, the listener used to verify events are sent only lives for 10 seconds, and then will exit as failed, then restart.

I believe this causes timing issues with this test, which leads to flakiness. Because the listener is failing/restarting every 10 seconds, there is increased chance that when the vspheresource attempts to send the events to the listener, there is nothing at the endpoint, and the events never make it to the listener. I want to adjust these tests to avoid this scenario.

## Proposed Changes

- :broom: Update or clean up current behavior
- this change makes the test more specific, and avoids timing problems
- the listener now expects events related to the actual job the test runs, namely, events signifying that the two vms indeed were turned off. Right now, it only checks on the type of the event 
- the listener no longer has a timeout. It will run until it receives the expected number of events. If it never receives them, the CancelFunc returned by RunJobListener will clean up the job+pod. By constantly being up, this will hopefully avoid timing issues that lead to flakes.

### Pre-review Checklist

- [ ] **At least 80% unit test coverage**
- [ ] **E2E tests** for any new behavior
- [ ] **Docs** for any user-facing impact

**Release Note**
N/A

### CODE REVIEW

File: test/test_images/listener/main.go
Comment (evankanderson) [inline test/test_images/listener/main.go:31]: Stylecheck wants "VM" capitalized per Go style

File: test/test_images/listener/main.go
Comment (evankanderson) [inline test/test_images/listener/main.go:62]: Instead of using `atomic`, I tend to collect the messages in a slice using channels, i.e.:

```go
events := make(chan cloudevents.Event, numExpectedEvents)
go func() {
  if err = client.StartReceiver(ctx, func(...) {
    if event.Type() == env.ExpectedEventType {
      events <- event
    }
    // Do rest of processing outside handler
  }); err != nil {
    // whinge and exit
  }
}()

for event := <-events {
  // Process events one by one, keeping count. When complete, call `cancel()` to exit StartReceiver, and break to end the loop.
}
```

Your way works as well, except for your cast from `int` to `int32` on line 79. One advantage to the channel approach is that you can handle each message serially and don't have to worry about mutexing any shared data structures, so you can handle the loop more "obviously". I'd also consider extracting the CloudEvent handler function so that you don't get a 10-line anonymous goroutine for the `go func() {...}()` construct.

**Review by evankanderson — COMMENTED** [review]
A few small nits, but I think this is overall a good improvement.

/assign @gabo1208 
/lgtm

**Comment by KauzClay** [comment]
hmm I can the run source test alone and it passes, but when I run all the tests together, it still fails sometimes. There must be something polluting across tests that I didn't realize....

File: test/test_images/listener/main.go
Comment (evankanderson) [inline test/test_images/listener/main.go:65]: Not sure if you want to replace the `logging.FromContext(ctx)` calls here and below with a single `log := logging.FromContext(ctx)` above and then just say `log.Fatalf(...)` through here.

File: test/test_images/listener/main.go
Comment (evankanderson) [inline test/test_images/listener/main.go:95]: This should be unreachable, right?

**Review by evankanderson — COMMENTED** [review]
Looks generally good -- are you ending up with not enough events being received?

If so, I'd check to make sure that the Trigger is `Ready` before you start sending the events. (and if so, there may be a bug!)

**Comment by embano1** [comment]
nit: there's a subtle bug in your event receiver channel logic which currently does not affect the receiver though.

If you receive more events than expected (for whatever reason), a net.poll goroutine in the cloudevents receiver handler could be blocked on sending to the channel while your main goroutine has already finished.

This is not an issue currently because you're not waiting for the cloudevents receiver to return. Just wanted to point this out here in case the code changes later which could cause a deadlock.

File: test/test_images/listener/main.go
Comment (KauzClay) [reply test/test_images/listener/main.go:65]: oh good point, I didn't realize that was an option 👍

File: test/e2e/source_test.go
Comment (KauzClay) [reply test/e2e/source_test.go:31]: I opted for hardcoding it because it seemed like a per test value. If we wrote a different test using the listener, I was thinking we would want to change the execpted event type and count.

Plus we can see that the event we are expecting is related to the actions we describe in the script further down (i.e. turning off vms), maybe making the test more clear?

File: test/test_images/listener/main.go
Comment (embano1) [reply test/test_images/listener/main.go:95]: I think what Evan meant is that this `if` branch would never be evaluated. I'd leave the `ctx.Cancel()` as is tbh.

**Comment by embano1** [comment]
Question @KauzClay : seeing a lot of flakes in other E2E runs, like this one (https://github.com/vmware-tanzu/sources-for-knative/runs/5884589946?check_suite_focus=true#step:7:123). Seems like success is not correctly handled due to an object creation/termination race perhaps. Wondering if your PR will fix this too?

**Comment by embano1** [comment]
> @embano1 I'm hoping it will. I've had trouble getting the tests to pass on this PR, so I tried bumping up the test timeout as you do in #414, and that helped TestSource pass.
> 
> 
> 
> But now it failed on the TestBindingPowerCLICore. I'll have to look into that, I haven't done much with that test so far

Check out the associated PR which also fixes the binding issue IMHO

**Comment by KauzClay** [comment]
@embano1 I see. Adding that got things to pass locally. However, now I'm back to having the TestSource fail again. Can't win with this one

**Comment by KauzClay** [comment]
Hey @embano1 , I've changed the tests to create the vcsim once, then run all the tests using the same vcsim. This seems to make things run much quicker.

I'm not very well versed in the go testing package, but I did the best I could at making the test do some setup prior to running all the tests. I took some inspiration from the [knative networking conformance tests](https://github.com/knative/networking/blob/main/test/conformance/ingress/run.go). 

At the moment, I couldn't find a good way to do it without always using the -run flag pointing to the "SuiteTest". This means you can't use the -run flag to run tests individually, so I tried adding some new extra flags.

If you or anyone else knows a nicer way to do this, I'm happy to change it. I do think creating only 1 vcsim for all the tests helps with flakiness though.

**Comment by embano1** [comment]
@KauzClay happy to debug this one further. Do we really need to introduce a `suite`? We did not need it before this PR so I think there's probably an easier way to fix the E2E tests.

**Comment by gabo1208** [comment]
@KauzClay I believe https://pkg.go.dev/testing#hdr-Main could help with your setup and teardown :)
As it is the PR is good for me (just some minor nits), but if it's not urgent, and you want to make the change I'm happy to wait (or merge otherwise)

**Comment by embano1** [comment]
Tbh, I'd rather have clean vcsim instances for each test scenario than reusing vcsim across the different tests. This would also keep our tests and the usage/doc simple.

This could hide potential timeouts and consistency issues in real-world deployments.

Can we investigate why setup/teardown of multiple vcsim instances is failing? I guess it's probably another race/timing issue (see earlier comment).

**Comment by gabo1208** [comment]
For the scenario @embano1 is suggesting I recommend a setup like [RabbitMQ's e2e tests](https://github.com/knative-sandbox/eventing-rabbitmq/blob/main/test/e2e/main_test.go)

You don't have to use the magic env and anything like that, but just call the setup function step before your test and then the teardown one after:
`func Test...(t *testing.T) {
	t.Parallel() // optional
	env.Test(ctx, t, SetupVCSim())
	env.Test(ctx, t, BlahTest())
	env.Test(ctx, t, TearDown())
}`

**Comment by embano1** [comment]
@KauzClay this should be good now, right? Can you please squash the commits?

**Comment by gabo1208** [comment]
Looking good to me, lets wait for @embano1 's approval

**Review by embano1 — APPROVED** [review]
AWESOME!

**Comment by embano1** [comment]
SHIP IT 🚀

**Review by gabo1208 — APPROVED** [review]

**Pull request merged**
